ecosystem/sep-0040.md

```diff
@@ -0,0 +1,198 @@
+  4. Any account holding that asset lays claim to the NFT.
+
+The NFT model involves two levels of indirection in order to minimize on-chain storage:
+
+    +---------------+      +----------+      +---------+
+    |  Stellar NFT  | ---> | metadata | ---> |   NFT   |
+    |  account data |      |   JSON   |      | content |
+    +---------------+      +----------+      +---------+
+
+
+## Specification
+This SEP defines an [issuance model](#issuance-model) that defines how NFT assets are issued on Stellar and a [metadata specification](#nft-metadata-specification) that defines how data about the NFT is captured. The metadata is stored in a way such that the Stellar network guarantee its authenticity.
+
+### Issuance Model
+Standard issuance of NFTs must take the following steps:
+
+  1. Create an account to represent the NFT: `CreateAccountOp(A, <reserve>)`
+  2. Configure the NFT's metadata pointer, as described [below](#nft-account-description): `ManageDataOp("type", "...")`
+  3. Form an asset describing your NFT: `Code:Issuer` (where issuer is `A`)
+  4. Create a distribution account: `B = CreateAccountOp()` and `ChangeTrust("Code:Issuer", <limit>)`
+  5. Issue the tokens: `PaymentOp(B, "Code:Issuer", <amount>)`
+  6. (Optional) Lock the issuing account: `SetOptionsOp(MasterWeight: 0)`
+
+Note that without this final step, there are concerns around both immutability and supply: an account may modify its NFT metadata pointer or issue more tokens unless it's locked down. Thus, Step 6 is recommended.
+
+Ownership of the token directly conveys ownership of a unit of the NFT, however that may be defined. Other scenarios such as ownership transfers via payments are built on top of this basic issuance model and are thus extraneous to this SEP.
+
+Other issuance models may exist and may be compatible with the following specification, but this is the standard issuance model on Stellar.
+
+### NFT Specification
+This specification is partially inspired by Ethereum's [EIP-721 specification](https://eips.ethereum.org/EIPS/eip-721#specification) on NFTs; however, it is generalized to arbitrary asset types (i.e. not necessarily images).
+
+In order to minimize the costs of on-chain storage, we have multiple levels of indirection of NFT metadata. The only thing the NFT account stores on-chain is a description of where to find the metadata.
+
+#### NFT Account Description
+A proper NFT account should specify its `homedomain`. It should also configure the asset in its domain's `stellar.toml` file (see [SEP-1](https://stellar.org/sep-1#currency-documentation)) if appropriate.
+
+There are two **required** fields on the account itself:
+
+| Field       | Purpose |
+| ----------- | ------- |
+| `type`      | Describes how the "first hop" of metadata is defined, standard values include `ipfshash` and `uri` |
+| `hash`      | The [SHA3](https://csrc.nist.gov/publications/detail/fips/202/final) hash of the metadata contents |
```

Comment:
I came here to chime in with the same thought. SHA3-256 is not an "upgraded" version of SHA2-256, it is considered an alternative should a break in the SHA2 family be found.

I would make the suggestion that perhaps there should be a 'hashType' field in addition to the 'hash' field. This would allow any arbitrary hash function type to be used (e.g. `sha`, `sha2-256`, `sha2-512`, `sha3-256`, `sha3-512`, `blake3`). If this approach is taken an enum of accepted types could be consulted to determine if what is being submitted is an acceptable type. I would continue to allow SHA1, even though it is not considered to be as secure, only to handle the case where content can be stored in a public Git repository and pointed to by the hash of its commit.

I would also suggest specifying the hash encoding as part of the spec. Hexadecimal encoding is the more popular way to represent hashes, but Base64 is also common and more compact. You should choose one and validate it on submission.

Cheers.

Reply:
Storing hashType assumes a single hash type and so I think we store different hashes types in unique fields and applications can simply look for their fields for the hash types they support. i.e. store the sha256 in a `sha256` field. If someone decided to store a sha-3 hash too, they can instead or in addition store a `sha3` field.

We don't need to encode the hashes, but should be explicit and clear that the hash is stored in it's binary form in the data value. Applications can render the hash in whatever encoding they find most appropriate.

Reply:
I would take these comments a step further and *require* that a hash of the content be provided in all cases. It has been a *major* shortcoming of existing NFT specs to allow linking to content that is not committed to by a hash of the content. It is so easy to hash the bytes of any content that it should never be skipped.

This can be the case even if IPFS is chosen as the `url`. An IPFS CID is not a file content hash and is harder for people to verify independently.

See:

https://docs.ipfs.io/concepts/hashing/#content-identifiers-are-not-file-hashes

By including a `hash` and `hashType` you make it much easier for client libs to verify the integrity of the file content using commonly found hashing libraries or stdlib functions. The IPFS hash is application specific code and its method of verification may change over time and should be considered a bonus integrity check on top of the content hash.

Reply:
Sha3 was supposed to (a) future-proof and (b) align with Ethereum, but I appreciate the benefit of sticking to what we already use.

Reply:
Side note. Ethereum uses Kechack256 which is not the same as SHA3-256. Different internal padding results in different hash outputs. So using SHA3 would not end up aligning with Ethereum in any case.

https://ethereum.stackexchange.com/questions/30369/difference-between-keccak256-and-sha3#30373

Reply:
Blegh, I hate that GitHub doesn't show me new comments as they come in. My comment was only in reply to leigh's first comment...

Reply:
Good thoughts here, thanks both! So we'll have an enumeration of possible keys that correlate to hashes? I don't really like that, and would prefer a type field. It makes it easier to look it up, no? Otherwise you have clients doing strange for-loops looking for a hash entry.

Regarding IPFS, this may be a product of my own ignorance, but I think CIDs still give you integrity, right? You know that the content at the CID is exactly the content put there by the issuer since if it changed, the CID would also change. There's no _need_ to verify that hashes match, since it's baked into the CID, right? To be clear, I do still think the hash is valuable for integrity (and especially in the non-IPFS case), I'm just trying to clarify my own understanding.